### PR TITLE
[SPARK-38069][SQL][SS] Improve the calculation of time window

### PR DESCRIPTION
**Pull request by nyingping**
### What changes were proposed in this pull request?

Remove the CaseWhen，Modified the calculation method of the obtained window

new logic：
`lastStart ` needs to be, which is less than `timestamp ` and is the maximum integer multiple of `windowsize`

`lastStart `is equal to `timestamp `minus the time left in the maximum integer multiple window
`val lastStart = timestamp - (timestamp - window.startTime
            + window.slideDuration) % window.slideDuration`

After getting `lastStart`, `lastEnd `is obvious, and other possible Windows can be computed using `i` and `windowsize`
### Why are the changes needed?

Structed Streaming computes window by intermediate result windowId, and windowId computes window by CaseWhen.

We can use Flink's method of calculating window to write it, which is more easy to understand, simple and efficient

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

Existing test as this is just refactoring. 

Also composed and ran a simple benchmark in this commit: https://github.com/HeartSaVioR/spark/commit/d532b6f6bcdd80cdaac520b21587ebb69ff2df8f

Quoting queries used to benchmark the change:

> tumble window

```
    spark.range(numOfRow)
      .selectExpr("CAST(id AS timestamp) AS time")
      .select(window(col("time"), "12 seconds", "12 seconds", "2 seconds"))
      .count()
```

> sliding window

```
    spark.range(numOfRow)
      .selectExpr("CAST(id AS timestamp) AS time")
      .select(window(col("time"), "17 seconds", "5 seconds", "2 seconds"))
      .count()
```

Results are following:

> tumble window

```
[info] OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~18.04-b10 on Linux 5.4.0-1045-aws
[info] Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
[info] tumbling windows:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] old logic                                            22             31          13        457.0           2.2       1.0X
[info] new logic                                            17             19           2        589.9           1.7       1.3X
```

> sliding window

```
[info] OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~18.04-b10 on Linux 5.4.0-1045-aws
[info] Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
[info] sliding windows:                          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] old logic                                          1347           1368          16          7.4         134.7       1.0X
[info] new logic                                           867            886          16         11.5          86.7       1.6X
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @hvanhovell FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -3884,14 +3884,12 @@ object TimeWindowing extends Rule[LogicalPlan] {
           case _ => Metadata.empty
         }
 
+
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3887]: nit. Please remove this redundant new line addition.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -3884,14 +3884,12 @@ object TimeWindowing extends Rule[LogicalPlan] {
...
         def getWindow(i: Int, overlappingWindows: Int, dataType: DataType): Expression = {
-          val division = (PreciseTimestampConversion(
-            window.timeColumn, dataType, LongType) - window.startTime) / window.slideDuration
-          val ceil = Ceil(division)
-          // if the division is equal to the ceiling, our record is the start of a window
-          val windowId = CaseWhen(Seq((ceil === division, ceil + 1)), Some(ceil))
-          val windowStart = (windowId + i - overlappingWindows) *
-            window.slideDuration + window.startTime
+          val timestamp = PreciseTimestampConversion(window.timeColumn, dataType, LongType)
+          val lastStart = timestamp- (timestamp- window.startTime
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3890]: Minior spacing issues: `timestamp-` -> `timestamp -` in two places.

Comment (nyingping) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: ok,thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (nyingping) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: > nit. Please remove this redundant new line addition.
@dongjoon-hyun 
ok,thanks!

**Comment by HeartSaVioR** [comment]
Thanks for the contribution! 

Given the code change is critical to the fixed time window calculation, could you please fill out the details on math in the PR description, in the section `What changes were proposed in this pull request?` ? 

It would be great if you can provide some calculation examples as well, tumble window / sliding window with start time.

Since the existing logic works for years, we need to be very confident on changing the logic even the new logic is considered as the better one. (We often struggle about regression, especially correctness issue.)

Thanks for understanding.

**Comment by HeartSaVioR** [comment]
cc. @brkyvz since he's an author of the code, although the code was committed 5+ years ago.

@nyingping 
If you don't mind, could you please try out micro-benchmark against the change?

Benchmarks in SQL are located in sql/core/test/scala, with package `org.apache.spark.sql.execution.benchmark`. Since it's really about creating time window, you don't need to deal with streaming query and aggregation. You can start with batch query (say, starting your Dataset via `spark.range(10000000)`) and convert these values to timestamp, and call `window` in select, and write to "noop" format of sink, and done.

Below is the simple benchmark code from #18364 - it didn't leverage the benchmark framework, but you can get some sense on creating benchmark code. In benchmark framework you'd like to remove spark.time and leverage the functionality of benchmark framework.

```
import org.apache.spark.sql.functions._

spark.time { 
  spark.range(numRecords)
    .select(from_unixtime((current_timestamp().cast("long") * 1000 + 'id / 1000) / 1000) as 'time)
    .select(window('time, "10 seconds"))
    .count()
}
```

If you feel too much bootstrapping on learning benchmark framework, please start with above code (with tumble/sliding) and if the code can show the difference, it would be sufficient.

**Comment by nyingping** [comment]
> 
@HeartSaVioR
I totally understand. Thank you very much.

I have modified the content according to your suggestion, and I am looking forward to your review

**Comment by HeartSaVioR** [comment]
https://github.com/HeartSaVioR/spark/commit/d532b6f6bcdd80cdaac520b21587ebb69ff2df8f

I just played with my own simple benchmark (in the commit above), and the gain is much more than the PR description. It's up to 30% for tumble window to 60% for sliding window. (I expect the gain gets bigger if maxNumOverlapping is higher.)

I'll update the PR description to contain the benchmark result.

I also did some calculations based on the new math to create sliding windows with offset by hand, and it seemed OK. I can't think of cases the new math may miss.

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @viirya @xuanyuanking Would like to have another eyes of reviewers. Thanks in advance!

**Comment by HeartSaVioR** [comment]
cc. @alex-balikov @jerrypeng as well.

**Comment by nyingping** [comment]
@HeartSaVioR 

Thank you for providing a more professional benchmark.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -3843,8 +3843,8 @@ object TimeWindowing extends Rule[LogicalPlan] {
    * The windows are calculated as below:
    * maxNumOverlapping <- ceil(windowDuration / slideDuration)
    * for (i <- 0 until maxNumOverlapping)
-   *   windowId <- ceil((timestamp - startTime) / slideDuration)
-   *   windowStart <- windowId * slideDuration + (i - maxNumOverlapping) * slideDuration + startTime
+   *   lastStart <- timestamp - (timestamp - startTime + windowDuration) % windowDuration
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3846]: Hmm, I suppose this is the window start time of last window. Don't we need to consider slide duration when calculating last start? I think this is only correct if the slide duration equals to window duration.

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3846]: That's inconsistency between code and comment :) Nice finding.

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3846]: @nyingping Could you please fix the comment here? Thanks!

**Review by viirya — COMMENTED** [review]
Seems okay, if I don't miss anything when verifying the calculation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3846]: @HeartSaVioR Oh, yea, I verified the calculation by looking at the comment. The code looks correct.

**Comment by nyingping** [comment]
remove parameter `maxNumOverlapping` from `getwindow` [on this commit](https://github.com/apache/spark/pull/35362/commits/625b4eb8a0faec5f7c0c519b49e8af54739bbb4d)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3846]: The code comment is not fixed; please don't resolve the conversation manually. You can fix it, push the commit and the comment thread will become outdated.

**Comment by nyingping** [comment]
@HeartSaVioR @viirya yes,of course.thanks!
Is this fix ok?

lastStart <- timestamp - (timestamp - startTime + windowDuration) % windowDuration
->
lastStart <- timestamp - (timestamp - startTime + slideDuration) % slideDuration


<< I do not know why this reply is always pending and cannot be displayed, so I reply here. Sorry

**Comment by viirya** [comment]
Yea, just to update the comment based on the code you changed, i.e. to make them consistent.

**Comment by nyingping** [comment]
> 

@HeartSaVioR @HeartSaVioR 
fixed comment and commit,thanks.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks. Merging to master.

**Comment by dongjoon-hyun** [comment]
Thank you all!

**Pull request closed without merging**